`Assessment`: Add shortcuts to save and submit an assessment

valeriebucher opened the pull request:
### Checklist
#### General
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x]  I added multiple screenshots/screencasts of my UI changes. 
- [x]  I translated all newly inserted strings into English and German. 

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Clicking the `save`, `submit` and `assess next submission` button are frequently used operations for tutors. To increase efficiency, a tutor should be able to save & submit their assessment as well as accessing the next submission using keyboard shortcuts.

### Description
<!-- Describe your changes in detail -->
They are displayed as tooltips above the respective button.

Shortcut for **saving** the assessment:
`Control` + `S` (Mac), `Ctrl` + `S` (Windows)

Shortcut for **submitting** and **overriding** the assessment:
`Control` + `Enter` (Mac), `Ctrl`+ `Enter`(Windows)

Shortcut for **assessing next submission**:
`Control` + `Shift` + `ArrowRight` (Mac), `Ctrl`+ `Shift` + `ArrowRight`(Windows)

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- 1 Tutor
- 1 Student submission of a text/modeling/file-upload/programming exercise(with manual assessment)

1. Navigate to the exercise dashboard
2. Start an assessment
3. Assess something and use the shortcut: `Ctrl` + `S` to save your assessment
4. Make sure your changes are saved
5. Submit your assessment by using the shortcut: `Ctrl`+ `Enter`
6. Make sure your assessment was submitted
7. Override the assessment by using the saving shortcut
8. Go to next submission by using the shortcut: `Ctrl` + `Shift` + `ArrowRight` (You will be informed in case there are no open submissions left)

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots
German version displays `Strg`

<img width="84" alt="Bildschirmfoto 2022-03-24 um 15 48 10" src="https://user-images.githubusercontent.com/73841445/159953421-d3bd9f33-0d54-4348-bfec-78a9f985f94b.png"> <img width="116" alt="Bildschirmfoto 2022-03-24 um 15 52 35" src="https://user-images.githubusercontent.com/73841445/159953432-367f799c-6405-43a5-a155-0c5a0cb4875f.png"> 
<img width="92" alt="Bildschirmfoto 2022-03-24 um 15 48 21" src="https://user-images.githubusercontent.com/73841445/159953427-9c87b2b6-3878-409d-88e2-d811f25f2587.png"> <img width="173" alt="Bildschirmfoto 2022-03-24 um 15 48 52" src="https://user-images.githubusercontent.com/73841445/159953430-61695fed-fe41-4f30-98b9-9e62d7cdad71.png"> <img width="193" alt="Bildschirmfoto 2022-03-24 um 15 48 37" src="https://user-images.githubusercontent.com/73841445/159953428-1d5be3cb-73b2-467b-b221-461147d8e13c.png">

Strohgelaender reviewed (commented):
Can you add the same functionality for File Upload and Programming Exercises?

valeriebucher commented:
@Strohgelaender I will do it within another PR before the start of the next semester.

ole-ve reviewed (dismissed):
Tested on TS5, works as described.
Code changes also look good to me 👍

b-fein commented on src/test/javascript/spec/component/modeling-assessment-editor/modeling-assessment-editor.component.spec.ts:
```diff
@@ -438,4 +438,23 @@ describe('ModelingAssessmentEditorComponent', () => {
+
+        expect(spyOnControlAndS).toHaveBeenCalledTimes(1);
+        expect(saveStub).toHaveBeenCalledTimes(1);
+    });
```
```suggestion
    });

```

b-fein reviewed (changes requested):
:+1: for adding keyboard shortcuts. The submit action works as expected.

However, the save shortcut also opens the ‘Save Webpage as HTML’ dialogue when using Firefox (on Linux, probably on Windows too).

Additionally, I think at the moment the user won’t ever know that those keyboard shortcuts exist. The UI should have a note somewhere about that (maybe as a tooltip each on the Save and Submit buttons?).

valeriebucher commented:
Thank you for testing and your feedback @Strohgelaender @ole-ve @b-fein 
The following has changed:
- Updated the method to suppress any default behavior (firefox)
- Added another shortcut (Ctrl N) to ‚assess next submission‘
- Added the functionality to programming and file upload exercises as well
- Tool tips are visible on each button (see screenshots)

Would appreciate it if you could review it - thank you!

b-fein commented on src/main/webapp/app/assessment/assessment-header/assessment-header.component.html:
```diff
@@ -144,7 +144,12 @@ <h3 class="top-container flex-wrap flex-lg-nowrap">
+                    class="btn m-1 btn-primary"
+                    (click)="save.emit()"
+                    [disabled]="!assessmentsAreValid || !isAssessor || saveBusy || submitBusy || cancelBusy"
+                    [ngbTooltip]="'Ctrl S'"
```
```suggestion
                    [ngbTooltip]="'Ctrl+S'"
```
The usual convention is that keyboard shortcuts are shown with `+` between the keys instead of spaces.

b-fein commented on src/main/webapp/app/assessment/assessment-header/assessment-header.component.html:
```diff
@@ -154,6 +159,7 @@ <h3 class="top-container flex-wrap flex-lg-nowrap">
                     class="btn m-1 btn-success"
                     (click)="submit.emit(); sendSubmitAssessmentEventToAnalytics()"
                     [disabled]="!assessmentsAreValid || !isAssessor || saveBusy || submitBusy || cancelBusy"
+                    [ngbTooltip]="'Ctrl Enter'"
```
```suggestion
                    [ngbTooltip]="'Ctrl+Enter'"
```

b-fein commented on src/main/webapp/app/assessment/assessment-header/assessment-header.component.html:
```diff
@@ -168,7 +174,13 @@ <h3 class="top-container flex-wrap flex-lg-nowrap">
+                *ngIf="result?.completionDate && canOverride"
+                (click)="submit.emit()"
+                [disabled]="!assessmentsAreValid || submitBusy"
+                [ngbTooltip]="'Ctrl Enter'"
```
```suggestion
                [ngbTooltip]="'Ctrl+Enter'"
```

b-fein commented on src/main/webapp/app/assessment/assessment-header/assessment-header.component.html:
```diff
@@ -187,6 +199,7 @@ <h3 class="top-container flex-wrap flex-lg-nowrap">
                 *ngIf="result?.completionDate && (isAssessor || exercise?.isAtLeastInstructor) && !hasComplaint && !isTeamMode && !isTestRun"
                 [disabled]="nextSubmissionBusy || submitBusy"
                 (click)="nextSubmission.emit(); sendAssessNextEventToAnalytics()"
+                [ngbTooltip]="'Ctrl N'"
```
```suggestion
                [ngbTooltip]="'Ctrl+N'"
```

b-fein reviewed (commented):
Code looks good.

The save and submit shortcuts worked for text and programming exercises, I assume they work for the other two exercise types as well as the code is pretty much the same. :+1: 

On Firefox, the Ctrl+N shortcut again is problematic, as it opens a new window. Firefox seems to catch this shortcut somewhere earlier already so the `event.preventDefault();` does not work.

Strohgelaender reviewed (changes requested):
I think we need to repeat the disabled-conditions of the buttons in their corresponding methods. As shown in the screenshot I can trigger the submit()-method using ctrl + Enter even if the button is disabled.
![grafik](https://user-images.githubusercontent.com/26540346/159029972-3f37a23a-8cac-44a2-9c24-a691903e2612.png)

valeriebucher commented:
@b-fein @Strohgelaender thanks for your feedback! I refactored it a bit and moved the hostlisteners directly to the assessment-header - I think this way is better because I avoid the code duplication in the files of the different exercise types, and also make sure that all conditions are met.

b-fein commented on src/test/javascript/spec/component/assessment-shared/assessment-header.component.spec.ts:
```diff
@@ -328,4 +328,82 @@ describe('AssessmentHeaderComponent', () => {
+        const spyOnControlAndS = jest.spyOn(component, 'saveOnControlAndS');
+        const saveSpy = jest.spyOn(saveButton, 'click');
+
+        expect(saveButton).toBeTruthy();
```
```suggestion
        expect(saveButton).not.toBeNull();
```
As per the guidelines `toBeTruthy/Falsy` shouldn’t be used as that can lead to accidentally wrong assertions (e.g., `0` is defined but also falsy).

b-fein reviewed (changes requested):
The save-shortcuts now still works when overriding assessments and no save button is shown. 

Otherwise, all shortcuts now work as intended without Firefox having other ideas on what those key combinations should do. :+1: 

I also think the code structure can be improved: Querying the button from the HTML and then ‘clicking’ it is likely to silently break if someone changes the id in the HTML without adapting the Typescript code. I added inline comments with a suggestions how this could be fixed for the save button by using the already existing event emitters. The same idea should be applicable to the other buttons as well.

valeriebucher commented on src/main/webapp/app/assessment/assessment-header/assessment-header.component.html:
```diff
@@ -144,7 +144,13 @@ <h3 class="top-container flex-wrap flex-lg-nowrap">
+                    id="save"
+                    class="btn m-1 btn-primary"
+                    (click)="save.emit()"
+                    [disabled]="!assessmentsAreValid || !isAssessor || saveBusy || submitBusy || cancelBusy"
```
Okay thanks! Yes but if someone changes the HTML, the client tests I added would fail, that's why I thought it should be okay, but your suggestion is better.👍

b-fein reviewed (approved):
All shortcuts now work as expected and perform no action when the corresponding button is disabled or not visible. :+1:

Code looks good, too.

willergeorg reviewed (approved):
Tested on TS1. The shortcuts work good. One thing I assume to be a bit misleading in your testing steps is step 8. If I used the shortcut to access the next submission (that was not present), the usual window occurred informing me that no open submissions left. This is of course not a deal breaker since this is the current delegation mechanism. But I understood your last testing step in that way that Artemis reroutes me to the exercise dashboard directly in this case. That was not the case for me.

willergeorg reviewed (approved):
Code looks good as well btw.

krusche commented on src/main/webapp/app/assessment/assessment-header/assessment-header.component.ts:
```diff
@@ -60,6 +61,7 @@ export class AssessmentHeaderComponent {
     // Icons
     faSpinner = faSpinner;
     faSave = faSave;
+    faArrowRight = faSquareCaretRight;
```
does not look good and might be confusing or is it on purpose?

valeriebucher replied:
It was on purpose, because the icon is used for the tooltip indicating that you can press the right arrow on the keyboard. But of course this might be confusing.

krusche replied:
I know the change was in purpose, but probably not the inconsistency that left and ride side are different? can you refactor (rename) the left side as well? I think the convention is that both sides are the same.
Otherwise, developers who only look into the html are confused why the icon looks different than the name would suggest

b-fein commented on src/main/webapp/app/assessment/assessment-header/assessment-header.component.html:
```diff
@@ -181,12 +193,16 @@ <h3 class="top-container flex-wrap flex-lg-nowrap">
                 <fa-icon *ngIf="submitBusy" [icon]="faSpinner" [spin]="true"></fa-icon>
                 <span jhiTranslate="artemisApp.assessment.button.useAsExampleSubmission">Use as Example Submission</span>
             </button>
+            <ng-template #nextSubmissionShortcut
+                >{{ 'artemisApp.assessment.button.control' | artemisTranslate }} + Shift + <fa-icon [icon]="faSquareCaretRight"></fa-icon
+            ></ng-template>
```
```suggestion
            <ng-template #nextSubmissionShortcut>
                {{ 'artemisApp.assessment.button.control' | artemisTranslate }} + Shift + <fa-icon [icon]="faSquareCaretRight"></fa-icon>
            </ng-template>
```
Looks like the automatic formatting placed the angles at the wrong places again.

b-fein reviewed (approved):
Changes since last review are only a renamed variable and code formatting. Looks good. :+1:

willergeorg reviewed (approved):
Changes since last review are nonbreaking. Reapprove.
